test/extensions/implementation/autograd.py

```diff
@@ -66,28 +66,49 @@ def variance(self):
         variances = [torch.var(g, dim=0, unbiased=False) for g in batch_grad]
         return variances
 
-    def diag_ggn(self):
-        _, output, loss = self.problem.forward_pass()
+    def _get_diag_ggn(self, loss, output):
+        diag_ggns = []
 
-        def extract_ith_element_of_diag_ggn(i, p):
+        def extract_ith_element_of_diag_ggn(i, p, loss, output):
             v = torch.zeros(p.numel()).to(self.problem.device)
             v[i] = 1.0
             vs = vector_to_parameter_list(v, [p])
             GGN_vs = ggn_vector_product_from_plist(loss, output, [p], vs)
             GGN_v = torch.cat([g.detach().view(-1) for g in GGN_vs])
             return GGN_v[i]
 
-        diag_ggns = []
         for p in list(self.problem.model.parameters()):
             diag_ggn_p = torch.zeros_like(p).view(-1)
 
             for parameter_index in range(p.numel()):
-                diag_value = extract_ith_element_of_diag_ggn(parameter_index, p)
+                diag_value = extract_ith_element_of_diag_ggn(
+                    parameter_index, p, loss, output
+                )
                 diag_ggn_p[parameter_index] = diag_value
 
             diag_ggns.append(diag_ggn_p.view(p.size()))
         return diag_ggns
 
+    def diag_ggn(self):
+        _, output, loss = self.problem.forward_pass()
+        return self._get_diag_ggn(loss, output)
+
+    def diag_ggn_batch(self):
+        batch_size = self.problem.input.shape[0]
+        _, _, batch_loss = self.problem.forward_pass()
+        batch_loss_device = batch_loss.device
+        loss_list = torch.zeros((batch_size), device=batch_loss_device)
```

Comment:
I think you can replace `(batch_size)` with `batch_size` and achieve the same

Reply:
I also saw that we have access to the `device` via `self.problem.device`. You can replace `batch_loss_device` and save one line.